### DISCUSSION TITLE
lowest night hours

### DISCUSSION THREAD

**Discussion by Nick75wien** [mampfes/ha_epex_spot#35] [Q&A]
Hello!

Thanks for the nice integration - I want to use it to trigger battery charge during low cost night hours.

Therefore I have adapted your calculation:

state: >- 
        {% set ns = namespace(attr_dict=[]) %}
        {% for item in (state_attr('sensor.epex_spot_at_price', 'data'))[0:24] %}
            {%- set ns.attr_dict = ns.attr_dict + [(loop.index-1,item["price_eur_per_mwh"])] %}
        {% endfor %}
        {%- set price_map = dict(ns.attr_dict) %}
        {%- set price_sort = price_map.values()|list %}
        {%- set keys_list = price_map.keys()|list %}
        {%- set ns = namespace(combo=[]) %}
        {%- for p in keys_list %}
          {%- set p = p|int %}
          {%- if p < 22 %}
            {%- set ns.combo = ns.combo + [(p, ((price_sort)[p] + (price_sort)[p+1] )|round(2))] %}
          {%- endif %}
        {%- endfor %}
        {%- set ns.combo = ns.combo[0:8] %}
        {%- set mapper = dict(ns.combo) %}
        {%- set key = mapper.keys()|list %}
        {%- set val = mapper.values()|list %}
        {%- set val_min = mapper.values()|min %}
        {{ key[val.index(val_min)]|string + ":00" }}        

But I need to have it as time to trigger the Automation with it - could you help me on that ?

KR Niki

**Comment by mampfes** [comment]
Sorry, I'm not very familiar with this syntax - the example was provided by an user.

Do you know the rank and the percentile sensor? Maybe they are sufficient for your request.

**Comment by Joebinator** [comment]
@Nick75wien 

If you want to run a simple automation then your trigger for the automation should look like this:

platform: template
value_template: >-
    {{ now().timestamp() | timestamp_custom('%H%M') == 
    today_at(states('sensor.your_sensor_name')).timestamp()
    | timestamp_custom('%H%M') }}
  
  I am running an automation like that to heat up my water boiler

**Reply by Nick75wien** [reply]
thanks for the hint - could you probably share your automation ?

**Reply by Joebinator** [reply]
sure =)

alias: Warmwasserheizung einschalten
description: ""
trigger:
  - platform: template
    value_template: >-
      {{ now().timestamp() | timestamp_custom('%H%M') == 
      today_at(states('sensor.beste_uhrzeit_warmwasserheizung_abend')).timestamp()
      | timestamp_custom('%H%M') }}
    id: ABEND
action:
  - choose:
      - conditions:
          - condition: trigger
            id:
              - ABEND
          - condition: template
            value_template: >-
              {{ states('sensor.warmwassertemperatur') | float  <=
              states('input_number.warmwasser_soll') | float -5 }}
        sequence:
          - service: script.turn_on
            data: {}
            target:
              entity_id: script.warmwasserheizung_einschalten
mode: single

**Reply by Nick75wien** [reply]
but how is that sensor defined ?
sensor.beste_uhrzeit_warmwasserheizung_abend

**Reply by Joebinator** [reply]
Just like in your first post, but just in case here is mine:

      - name: beste_uhrzeit_warmwasserheizung_abend
        unique_id: "Uhrzeit_warmwasseraufheizen"
        state: >- 
          {% set ns = namespace(attr_dict=[]) %}
          {% for item in (state_attr('sensor.epex_spot_at_price', 'data'))[0:24] %}
              {%- set ns.attr_dict = ns.attr_dict + [(loop.index-1,item["price_eur_per_mwh"])] %}
          {% endfor %}
          {%- set price_map = dict(ns.attr_dict) %}
          {%- set price_sort = price_map.values()|list %}
          {%- set keys_list = price_map.keys()|list %}
          {%- set ns = namespace(combo=[]) %}
          {%- for p in keys_list %}
            {%- set p = p|int %}
            {%- if p < 22 %}
              {%- set ns.combo = ns.combo + [(p, ((price_sort)[p])|round(2))] %}
            {%- endif %}
          {%- endfor %}
          {%- set ns.combo = ns.combo[0:24] %}
          {%- set mapper = dict(ns.combo) %}
          {%- set key = mapper.keys()|list %}
          {%- set val = mapper.values()|list %}
          {%- set val_min = mapper.values()|min %}
          {{ key[val.index(val_min)]|string + ":00" }}
          
          
That gives you the Time (1 Hour) for the lowest Price before midnight.

if you want the best time for exaplme 3 hours  of runtime (like in the documentation) then you need to change this section
{%- set ns.combo = ns.combo + [(p, ((price_sort)[p])|round(2))] %}

to

{%- set ns.combo = ns.combo + [(p, ((price_sort)[p] + (price_sort)[p+1] + (price_sort)[p+2])|round(2))] %}

**Reply by Nick75wien** [reply]
this actually gives 13:00

**Reply by Joebinator** [reply]
Yeah, because 13:00 is the cheapest hour today.

**Reply by Nick75wien** [reply]
ok! thats what I also use for starting the washing machine

but I now also have created a cheap night sensor to recharge my battery to have cheap electricity in the morning before the solar panels start working ;-)

thanks anyway for your help

**Comment by Hankersern** [comment]
Hey there, maybe you could help me.
My battery should be loaded at lowest prices and needs about 3 hours to charge. 

But the sensor seems to be wrong. It should be 13:00 but show 01:00.

![screenshot_battery_lowperiod](https://github.com/mampfes/ha_epex_spot/assets/86988609/f7463693-75d5-43cc-b954-050e275c5ac1)

edit - my sensor:
```
## Sensor guenstigster Zeitpunkt um Batterien zu laden ##
template:
  - sensor:
      - name: epex_start_low_period
        state: >-
          {% set ns = namespace(attr_dict=[]) %}
          {% for item in (state_attr('sensor.epex_spot_de_lu_price', 'data'))[0:24] %}
              {%- set ns.attr_dict = ns.attr_dict + [(loop.index-1,item["price_eur_per_mwh"])] %}
          {% endfor %}
          {%- set price_map = dict(ns.attr_dict) %}
          {%- set price_sort = price_map.values()|list %}
          {%- set keys_list = price_map.keys()|list %}
          {%- set ns = namespace(combo=[]) %}
          {%- for p in keys_list %}
            {%- set p = p|int %}
            {%- if p < 22 %}
              {%- set ns.combo = ns.combo + [(p, ((price_sort)[p] + (price_sort)[p+1] + (price_sort)[p+2])|round(2))] %}
            {%- endif %}
          {%- endfor %}
          {%- set mapper = dict(ns.combo) %}
          {%- set key = mapper.keys()|list %}
          {%- set val = mapper.values()|list %}
          {%- set val_min = mapper.values()|min %}
          {{ key[val.index(val_min)]|string + ":00" }}
```

Thanks!

**Reply by Joebinator** [reply]
Hi!

When i copy it into my home assistant template engine it gives me 13:00 maybe you have some different settings for time/date in your installation?

**Reply by Joebinator** [reply]
<img width="1230" alt="image" src="https://github.com/mampfes/ha_epex_spot/assets/44570824/7f9e8d77-2f50-40df-b90b-0a048c709aeb">

only thing i changed was sensor.epex_spot_de_lu_price to sensor.epex_spot_at_price since i do not have yours

**Reply by Hankersern** [reply]
Hey Joe,

thanks for your reply!
Interesting it works for you...
Unfortunately at the moment the sensor doesnt deliver any data (german internet is still neuland for the government) - this brings me to another idea: 

My Epex Integration gets its data from the german Site Smard.de - where do you grab the data? 

Beste Grüße :-)

**Reply by Joebinator** [reply]
I take my data from EPEX Spot Web Scraper

**Reply by Hankersern** [reply]
Ok, i'll change my integration to use the scraper. Let's see if this works :-)

**Reply by Hankersern** [reply]
Well, maybe the Scraper works better. But I have to keep a look on it - Screenshot should speak for itself. 
I mean, the best hours should be from ~4 to 7 and not 13:00h now, shouldnt? 

![screenshot_battery_lowperiod2](https://github.com/mampfes/ha_epex_spot/assets/86988609/1e9b17d5-c571-4e18-bb0a-55171885254c)

**Reply by Joebinator** [reply]
The time value you get is for **_today_**, after midnight it will change accordingly.

if anybody has an idea how to configure this template to look for the whole available data, i would also be very grateful =)

**Reply by Hankersern** [reply]
I will stay awake until 0:30 today to see if it changes :-D

**Reply by Joebinator** [reply]
don't ask my why, but mine changes always at 01:00

**Comment by Hankersern** [comment]
> don't ask my why, but mine changes always at 01:00

Epex releases the new prices daily at 13:00 oO

Mine seems to work so far. I'll change the automations for charging now and keep an eye on it.
Thanks again!


![screenshot_battery_lowperiod3](https://github.com/mampfes/ha_epex_spot/assets/86988609/26710059-0b79-4074-859e-694c4b28bf7e)
